### PR TITLE
Support for the binary distance matrix type produced by biocore/unifrac

### PR DESCRIPTION
**Pull request by wasade**
This code contribution provides support for the HDF5-based distance matrix type produced by `biocore/unifrac`. These changes were written in collaboration with @gwarmstrong 

Please complete the following checklist:

* [x] I have read the guidelines in [CONTRIBUTING.md](https://github.com/biocore/scikit-bio/blob/master/CONTRIBUTING.md).

* [x] I have documented all public-facing changes in [CHANGELOG.md](https://github.com/biocore/scikit-bio/blob/master/CHANGELOG.md).

* [ ] **This pull request includes code, documentation, or other content derived from external source(s).** If this is the case, ensure the external source's license is compatible with scikit-bio's [license](https://github.com/biocore/scikit-bio/blob/master/COPYING.txt). Include the license in the `licenses` directory and add a comment in the code giving proper attribution. Ensure any other requirements set forth by the license and/or author are satisfied. **It is your responsibility to disclose code, documentation, or other content derived from external source(s).** If you have questions about whether something can be included in the project or how to give proper attribution, include those questions in your pull request and a reviewer will assist you.

* [x] **This pull request does not include code, documentation, or other content derived from external source(s).**

**Note:** [REVIEWING.md](https://github.com/biocore/scikit-bio/blob/master/REVIEWING.md) may also be helpful to see some of the things code reviewers will be verifying when reviewing your pull request.

### CODE REVIEW

File: skbio/io/format/binary_dm.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+def _h5py_mat_to_skbio_mat(cls, fh):
+    return cls(fh['matrix'], _parse_ids(fh['order']))
+
+
+def _skbio_mat_to_h5py_mat(obj, fh):
+    _set_header(fh)
+
+    ids = fh.create_dataset('order', shape=(len(obj.ids), ), dtype=_vlen_dtype)
+    ids[:] = obj.ids
+    fh.create_dataset('matrix', data=obj.data)
```
Comment (gwarmstrong) [inline skbio/io/format/binary_dm.py:141]: Here it looks like the usage of `_skbio_mat_to_h5py_mat(fh)` in the `binary_dm` writers does not match the signature `_skbio_mat_to_h5py_mat(obj, fh)`

File: skbio/io/format/binary_dm.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
++-----------+---------------------+------------------------------+
+|Attribute  |Value                |Description                   |
+|           |type                 |                              |
++===========+=====================+==============================+
+|format     |string               |A string identifying the file |
+|           |                     |as Binary DM format           |
++-----------+---------------------+------------------------------+
+|version    |string               |The version of the current    |
+|           |                     |Binary DM format              |
```
Comment (gwarmstrong) [inline skbio/io/format/binary_dm.py:46]: How do you feel about the discordance between these being considered "required", but the sniffer below not checking for them?

File: skbio/stats/distance/tests/test_base.py [hunk 3]
```diff
@@ -1102,5 +1103,17 @@ def test_run_monte_carlo_stats_invalid_permutations(self):
...
+class DissimilarityMatrixTests(DissimilarityMatrixTestBase, TestCase):
+    matobj = DissimilarityMatrix
+    def setUp(self):
+        super(DissimilarityMatrixTests, self).setUp()
+
+
+class DistanceMatrixTests(DistanceMatrixTestBase, TestCase):
+    matobj = DistanceMatrix
+    def setUp(self):
+        super(DistanceMatrixTests, self).setUp()
```
Comment (gwarmstrong) [inline skbio/stats/distance/tests/test_base.py:1115]: Nice!

File: skbio/io/format/binary_dm.py (hunk 1 above)
Comment (wasade) [reply skbio/io/format/binary_dm.py:141]: ah, good catch!

File: skbio/io/format/binary_dm.py (hunk 2 above)
Comment (wasade) [reply skbio/io/format/binary_dm.py:46]: I think this is incorrect. The sniffer issues a call to `_get_header`, which in turn tests these are present in the hdf5 file.

Comment (gwarmstrong) [reply skbio/io/format/binary_dm.py:46]: Ah yep you're right here. Just did not see the keys there, got it now.

**Comment by wasade** [comment]
This is passing travis, but failing coveralls despite "Coverage remained the same". I think this may be good for final review?

cc @gwarmstrong @ElDeveloper

**Comment by ElDeveloper** [comment]
If that's OK with everyone, I think @ebolyen might be a better reviewer for this PR. Plus adding h5py as a dependency is something worthwhile discussing as a group.

**Pull request merged**